core/modules/replica/ControllerApp.cc

```diff
@@ -95,6 +98,42 @@ void printRequest(Request::Ptr const& request,
 }
 
 
+void printRequest(IndexRequest::Ptr const& request, string const& fileName) {
+    auto&& responseData = request->responseData();
+    cout << request->context() << "\n"
+         << "responseData.error: " << responseData.error << "\n"
+         << "performance:  " << request->performance() << endl;
+
+    if (fileName.empty()) cout << responseData.data;
+    else {
+        ofstream f(fileName);
+        if (not f.good()) {
+            cerr << "failed to open file: " << fileName << endl;
+            return;
+        }
+        f << responseData.data;
+    }
+}
+
+
+template <>
+void printRequest<StatusIndexRequest>(StatusIndexRequest::Ptr const& request) {
+    auto&& responseData = request->responseData();
+    cout << request->context() << "\n"
+         << "responseData.error: " << responseData.error << "\n"
+         << "performance:  " << request->performance() << endl;
```

Comment:
Having lots of variations of a templated printRequest function feels inside-out to me, why not add a virtual function (maybe pure?) to the base class & implement the specific variations of the function in the subclasses?

Reply:
This is an interesting problem. I'm using templates functions to extend classes. So, in theory, that `printRequest` functions could be virtual methods of the classes whose state is bing inspected by the current code. There were two reasons why I didn't do this when I was facing a similar dilemma. These were:
* I think the function templates produce less boiler-plate code compared with the virtual methods simply because most request classes would have similar implementations
* and the current class seems to be the only client of these functions. If I had more than one client then moving the functions elsewhere would be a good alternative.
Having said that, I'm still open to further refactoring of this code. I may not just do this today. My "rule of thumb" is that when I'm starting making improvements to some code and when I see something which makes it difficult to understand or modify the code then I go for a refactoring. Some sort of "lazy improvement" :smh